### PR TITLE
Speed-up time spent in generating shaped str keys

### PR DESCRIPTION
**Pull request by jhavukainen**
Replaces the janky way of using the IntArrayRef to create an NSArray to ask for it to provide its contents in a string format with use of stringstream. 

This speeds up the call for getting the key string for caching (or reading from cache) for shaped inputs by ~5x. While the actual wall time, depending on the number of input tensors, is only some microseconds this time represents non-negligible chunk of the overall time spent in preparing to dispatch work to the GPU. And since this function gets called on every time a (cacheable) operation in MPS is used it should be a small but broadly impacting time saver.

Using mps_linear as an example. Note this is before PR https://github.com/pytorch/pytorch/pull/152199 so it only captures the CPU time spent in the op call:

Before the change:
```
torch.linear time: <torch.utils.benchmark.utils.common.Measurement object at 0x1108f07d0>
func(*args, **kwargs)
  Median: 22.75 us
  IQR:    0.87 us (22.50 to 23.38)
  8361 measurements, 1 runs per measurement, 1 thread
```

After the change:
```
torch.linear time: <torch.utils.benchmark.utils.common.Measurement object at 0x108875350>
func(*args, **kwargs)
  Median: 18.67 us
  IQR:    0.46 us (18.50 to 18.96)
  10342 measurements, 1 runs per measurement, 1 thread
```

Which aligns with the observed change for getTensorStringKeys() taking ~1us instead of ~5us  in mps_linear op I got from a point measurement sandwiching the function call with `std::chrono::high_resolution_clock`.

cc @kulinseth @albanD @malfet @DenisVieriu97

### CODE REVIEW

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 1]
```diff
@@ -311,7 +311,12 @@ MPSDataType getMPSScalarType(ScalarType scalar_type) {
 std::string getArrayRefString(const IntArrayRef s) {
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/OperationUtils.mm:311]: This entire func can probably speed up a lot with libfmt.

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 2]
```diff
@@ -311,7 +311,12 @@ MPSDataType getMPSScalarType(ScalarType scalar_type) {
...
   std::stringstream ss;
   std::copy(s.begin(), s.end(), std::ostream_iterator<int>(ss, ","));
-  return ss.str();
+
+  // Remove spurious "," at the end from osstream_iterator.
+  auto refstring = ss.str();
+  refstring.pop_back();
+
+  return refstring;
 }
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/OperationUtils.mm:320]:
```suggestion
std::string getArrayRefString(const IntArrayRef s) {
  return fmt::format("{}", fmt::join(s, ","));
}

Comment (Skylion007) [reply aten/src/ATen/native/mps/OperationUtils.mm:320]: Includes required:
```cpp
#include <fmt/format.h>
#include <fmt/ostream.h> // only needed if you want to support overloading << of arguments`
#include <fmt/ranges.h> 
```

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 3]
```diff
@@ -311,7 +311,12 @@ MPSDataType getMPSScalarType(ScalarType scalar_type) {
...
 
 std::string getTensorsStringKey(const TensorList& tensors, bool short_dtype, bool exclude_shape) {
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/OperationUtils.mm:322]: Couldn't add diff for this but:
```cpp
std::string getTensorsStringKey(const TensorList& tensors, bool short_dtype, bool exclude_shape) {
  fmt::memory_buffer buffer;

  for (const Tensor& tensor : tensors) {
    fmt::format_to(buffer, ":");
    if (tensor.defined()) {
      fmt::format_to(buffer, "{}[", getMPSTypeString(tensor.scalar_type(), short_dtype));
      if (tensor.dim() == 0) {
        fmt::format_to(buffer, "Scalar");
      } else {
        if (exclude_shape) {
          fmt::format_to(buffer, "-1");
        } else {
          fmt::format_to(buffer, "{}", getArrayRefString(tensor.sizes()));
        }
      }
      fmt::format_to(buffer, "]");
      if (tensor.is_conj()) {
        fmt::format_to(buffer, "_conj");
      }
    } else {
      fmt::format_to(buffer, "Undefined");
    }
  }

  return fmt::to_string(buffer);
```

Comment (Skylion007) [reply aten/src/ATen/native/mps/OperationUtils.mm:322]: You can even do  `fmt::basic_memory_buffer<char, kSmallBufferSize>` buffer; where `kSmallBufferSize` is some number of chars that will be preallocated to the stack to prevent any heap calls except for the to_string. call

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 4]
```diff
@@ -328,8 +333,7 @@ MPSDataType getMPSScalarType(ScalarType scalar_type) {
         if (exclude_shape) {
           str += "-1";
         } else {
-          str +=
-              std::string([[getMPSShape(tensor) valueForKey:@"description"] componentsJoinedByString:@","].UTF8String);
+          str += getArrayRefString(tensor.sizes());
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/OperationUtils.mm:336]: Alternatively, just:
```suggestion
          fmt::format_to(str, fmt::join(s, ","));
```

File: aten/src/ATen/native/mps/OperationUtils.mm (hunk 3 above)
Comment (jhavukainen) [reply aten/src/ATen/native/mps/OperationUtils.mm:316]: Thanks @Skylion007 for the suggestions! Went ahead and changed and changed the it approx. to the imlementation you suggested. I did have to make the first input to `fmt::format_to` into a `std::back_inserter` based on brief reading of the docs as the API did not seem to accept the memory buffer directly. Might be some version difference? In any case feel free to add further comments if it looks off to you, I'm not too well versed in the finer details of C++ string formatting scene so I'm grateful for the comments.

After the changes repeating the above measurements gives me

```
torch.linear time: <torch.utils.benchmark.utils.common.Measurement object at 0x13b88c490>
func(*args, **kwargs)
  Median: 17.83 us
  IQR:    0.33 us (17.71 to 18.04)
  10819 measurements, 1 runs per measurement, 1 thread
```

And point measurement around the calls seems to land it somewhere in the < 0.5 us region which I think is probably good enough for now and would require a more rigorous measuring setup to get reliable results on squeezing it much further.

**Review by Skylion007 — APPROVED** [review]

File: aten/src/ATen/native/mps/OperationUtils.mm [hunk 5]
```diff
@@ -309,37 +310,35 @@ MPSDataType getMPSScalarType(ScalarType scalar_type) {
...
-  std::copy(s.begin(), s.end(), std::ostream_iterator<int>(ss, ","));
-  return ss.str();
+  return fmt::to_string(fmt::join(s, ","));
 }
 
 std::string getTensorsStringKey(const TensorList& tensors, bool short_dtype, bool exclude_shape) {
-  std::string str;
-  // The key format per tensor would look like ":Float32[1,1,1,10]:"
+  fmt::basic_memory_buffer<char, 100> buffer;
+  auto buf_iterator = std::back_inserter(buffer);
```
Comment (Skylion007) [inline aten/src/ATen/native/mps/OperationUtils.mm:318]: @cyyever do you find this buf_iterator problematic?
